client/ClientCommandManager.cpp

```diff
@@ -0,0 +1,540 @@
+				neutralAI->String() = fname;
+				printCommandMessage("Setting changed, from now the battle ai will be " + fname + "!\n");
+			}
+		}
+		catch(std::exception &e)
+		{
+			printCommandMessage("Failed opening " + fname + ": " + e.what(), ELogLevel::WARN);
+			printCommandMessage("Setting not changed, AI not found or invalid!", ELogLevel::WARN);
+		}
+	}
+	else if(commandName == "autoskip")
+	{
+		Settings session = settings.write["session"];
+		session["autoSkip"].Bool() = !session["autoSkip"].Bool();
+	}
+	else if(commandName == "gosolo")
+	{
+		ClientCommandManager::handleGoSolo();
+	}
+	else if(commandName == "controlai")
+	{
+		std::string colorName;
+		singleWordBuffer >> colorName;
+		boost::to_lower(colorName);
+
+		ClientCommandManager::handleControlAi(colorName);
+	}
+	else
+	{
+		printCommandMessage("Command not found :(", ELogLevel::ERROR);
+	}
+}
+
+void ClientCommandManager::giveTurn(const PlayerColor &colorIdentifier)
+{
+	YourTurn yt;
+	yt.player = colorIdentifier;
+	yt.daysWithoutCastle = CSH->client->getPlayerState(colorIdentifier)->daysWithoutCastle;
+	yt.applyCl(CSH->client);
+}
+
+void ClientCommandManager::removeGUI()
+{
+	// CClient::endGame
+	GH.curInt = nullptr;
+	if(GH.topInt())
+		GH.topInt()->deactivate();
+	GH.listInt.clear();
+	GH.objsToBlit.clear();
+	GH.statusbar = nullptr;
+	printCommandMessage("Removed GUI.", ELogLevel::INFO);
+
+	LOCPLINT = nullptr;
+}
+
+void ClientCommandManager::printInfoAboutInterfaceObject(const CIntObject *obj, int level)
+{
+	std::stringstream sbuffer;
+	sbuffer << std::string(level, '\t');
+
+	sbuffer << typeid(*obj).name() << " *** ";
+	if (obj->active)
+	{
+#define PRINT(check, text) if (obj->active & CIntObject::check) sbuffer << text
+		PRINT(LCLICK, 'L');
+		PRINT(RCLICK, 'R');
+		PRINT(HOVER, 'H');
+		PRINT(MOVE, 'M');
+		PRINT(KEYBOARD, 'K');
+		PRINT(TIME, 'T');
+		PRINT(GENERAL, 'A');
+		PRINT(WHEEL, 'W');
+		PRINT(DOUBLECLICK, 'D');
+#undef  PRINT
+	}
+	else
+		sbuffer << "inactive";
+	sbuffer << " at " << obj->pos.x <<"x"<< obj->pos.y;
+	sbuffer << " (" << obj->pos.w <<"x"<< obj->pos.h << ")";
+	printCommandMessage(sbuffer.str(), ELogLevel::INFO);
+
+	for(const CIntObject *child : obj->children)
+		printInfoAboutInterfaceObject(child, level+1);
+}
+
+void ClientCommandManager::printCommandMessage(const std::string &commandMessage, ELogLevel::ELogLevel messageType)
+{
+	switch(messageType)
+	{
+		case ELogLevel::NOT_SET:
+			std::cout << commandMessage;
+			break;
+		case ELogLevel::TRACE:
+			logGlobal->trace(commandMessage);
+			break;
+		case ELogLevel::DEBUG:
+			logGlobal->debug(commandMessage);
+			break;
+		case ELogLevel::INFO:
+			logGlobal->info(commandMessage);
+			break;
+		case ELogLevel::WARN:
+			logGlobal->warn(commandMessage);
+			break;
+		case ELogLevel::ERROR:
+			logGlobal->error(commandMessage);
+			break;
+		default:
+			std::cout << commandMessage;
+			break;
+	}
+
+	if(currentCallFromIngameConsole)
+	{
+		if(LOCPLINT && LOCPLINT->cingconsole)
```

Comment:
Access to GUI objects from another thread requires lock.
I think this is correct way to lock, but please recheck
```suggestion
		boost::unique_lock<boost::recursive_mutex> un(*CPlayerInterface::pim);
		if(LOCPLINT && LOCPLINT->cingconsole)
```